pyhealth/datasets/base_kg_dataset.py

```diff
@@ -0,0 +1,157 @@
+
+
+logger = logging.getLogger(__name__)
+
+INFO_MSG = """
+dataset.triples:
+    Array((<head_entity>, <relation>, <tail_entity>))
+"""
+
+
+class BaseKGDataset(ABC):
+    """Abstract base Knowledge Graph class
+    
+    This abstract class defines a uniform
+
+    Each specific dataset will be a subclass of this abstract class, which can then
+    be converted to samples dataset for different tasks by calling `self.set_task()`.
+
+    Args:
+        dataset_name: name of the dataset.
+        root: root directory of the raw data (should contain many csv files).
+        dev: whether to enable dev mode (only use a small subset of the data).
+            Default is False.
+        refresh_cache: whether to refresh the cache; if true, the dataset will
+            be processed from scratch and the cache will be updated. Default is False.
+    
+    """
+
+    def __init__(
+        self,
+        root: str,
+        dataset_name: Optional[str] = None,
+        dev: bool = False,
+        refresh_cache: bool = False
+    ):
+        self.dataset_name = (
+            self.__class__.__name__ if dataset_name is None else dataset_name
+        )
+        self.root = root
+        self.dev = dev
+        self.triples = []
+        self.samples = []
+        self.task_name = "Null"
+        self.entity_num = 0
+        self.relation_num = 0
+        self.refresh_cache = refresh_cache
+
+        # hash filename for cache
+        args_to_hash = [self.dataset_name, root] + ["dev" if dev else "prod"]
+        filename = hash_str("+".join([str(arg) for arg in args_to_hash]))
+        self.filepath = os.path.join(MODULE_CACHE_PATH, filename)
+        self.raw_graph_process()
+
+        self.len = len(self.triples)
+
+
+    def __str__(self):
+        """Prints some information of the dataset."""
+        return f"Base dataset {self.dataset_name}"
+
+    def __len__(self):
+        return self.len
+
+    def raw_graph_process(self):
+        """Process the raw graph to triples (a list of triple)
+        """
+        raise NotImplementedError
+
+    @staticmethod
+    def info():
+        """Prints the output format."""
+        print(INFO_MSG)
+
+    def stat(self):
+        """Returns some statistics of the base dataset."""
+        lines = list()
+        lines.append("")
+        lines.append(f"Statistics of base dataset (dev={self.dev}):")
+        lines.append(f"\t- Dataset: {self.dataset_name}")
+        lines.append(f"\t- Number of triples: {len(self.triples)}")
+        lines.append(f"\t- Number of entities: {self.entity_num}")
+        lines.append(f"\t- Number of relations: {self.relation_num}")
+        lines.append(f"\t- Task name: {self.task_name}")
+        lines.append(f"\t- Number of samples: {len(self.samples)}")
+        lines.append("")
+        print("\n".join(lines))
+        return 
+
+    
+    def set_task(
+        self,
+        task_fn: Callable,
+        task_name: Optional[str] = None,
+        save: bool = True,
```

Comment:
delete save. If refresh_cache = False in the dataset, then the task dataset will be saved automatically. Otherwise, it will not be saved.